kpatch-build/create-diff-object.c

```diff
@@ -1804,6 +1804,71 @@ static void kpatch_include_symbol(struct symbol *sym)
 		kpatch_include_section(sym->sec);
 }
 
+/*
+ * Module namespaces are saved as key value pairs in the .modinfo
+ * section.  For modules, they are in the form "import_ns=<value" and
+ * for built-ins, "<object>.import_ns=<value>".  Extract just these
+ * namespace strings from the patched .modinfo to put back into the
+ * output ELF.
+ */
+static void kpatch_get_modinfo_namespaces(struct kpatch_elf *kelf_patched,
+					  struct section *sec)
+{
+	char *data = NULL;
+	int data_len = 0;
+	const char *p, *end;
+
+	if (!sec->data || !sec->data->d_buf)
+		return;
+
+       /*
+        * Copy all the "import_ns=<value>" strings into
+        * a temporary data array (ignore other key-values)
+        */
+	p = sec->data->d_buf;
+	end = p + sec->sh.sh_size;
+
+	while (p < end) {
+		const char *match;
+		char *new_data;
+		size_t match_len;
+
+		match = memmem(p, end - p, "import_ns=", 10);
+		if (!match)
+			break;
+		match_len = strlen(match) + 1;
+
+		new_data = realloc(data, data_len + match_len);
+		if (!new_data)
+			ERROR("realloc");
+		data = new_data;
+
+		memcpy(data + data_len, match, match_len);
+		data_len += (int) match_len;
+
+		p = match + match_len;
+	}
+
+	/*
+	 * Update the patched elf modinfo data array with namespace
+	 * strings for kpatch_put_modinfo_namespaces()
+	 */
+	kelf_patched->modinfo_data = data;
+	kelf_patched->modinfo_data_len = data_len;
+}
+
+static void kpatch_put_modinfo_namespaces(struct kpatch_elf *kelf_patched,
+					  struct kpatch_elf *kelf_out)
+{
+	struct section *sec;
+
+	sec = create_section(kelf_out, ".modinfo",
```

Comment:
Should this be skipped if .modinfo doesn't exist in the patched object?

Reply:
Yup, no need to create an empty section.

Reply:
True, that's a better test, as it would skip instances with .modinfo but no imports.